test: Prove+document ConstevalFormatString/tinyformat parity

hodlinator opened the pull request:
Clarifies and puts the extent of parity under test.

Broken out from #30546 based on https://github.com/bitcoin/bitcoin/pull/30546#discussion_r1755013263 and https://github.com/bitcoin/bitcoin/pull/30546#discussion_r1756495304.

hodlinator commented on src/test/util_string_tests.cpp:
```diff
@@ -13,17 +13,50 @@ BOOST_AUTO_TEST_SUITE(util_string_tests)
-    // This was already executed at compile-time, but is executed again at run-time to avoid -Wunused.
-    decltype(fmt)::Detail_CheckNumFormatSpecifiers(fmt.fmt);
+    // Prove parity with tinyformat
+    switch (NumArgs) {
+    case 0:
```
Not sure it would reduce verbosity that much, but would definitely consider switching if someone took the time to provide a diff.

Compiler warnings are triggered for missing cases.

hodlinator replied:
While posting this PR I was aware of your `tuple_cat` suggestion. I am not against it on principle but TBH I haven't had a clear reason to struggle through that corner of C++ and don't see your suggestion compelling enough to work through it. Your suggestion in that comment is close to compiling but doesn't handle sending in one arg less for the failure cases. A motivation behind this PR is to prove parity in a clear way, and I'm worried it might obfuscate it. Happy to learn more from a diff.

Added default cases in dbbc2e52284a151a33d553b4a4032503906a4774, agree it's more robust.

hodlinator replied:
Gave it another go in the latest push (071342d4c7627200323d8f22cc37ea022b2e9d69) and your direction seems to be working fine. Happy to see the linter changes go!

hodlinator commented on test/lint/run-lint-format-strings.py:
```diff
@@ -15,6 +15,12 @@
+    ("src/test/util_string_tests.cpp", 'tfm::format(fmt.fmt, "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11")'),
+    ("src/test/util_string_tests.cpp", 'tfm::format(fmt.fmt, "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12")'),
+    ("src/test/util_string_tests.cpp", "tfm::format(fmt.fmt, a, b)"),
+    ("src/test/util_string_tests.cpp", "tfm::format(fmt.fmt, a, b, c)"),
```
I'll be happy to remove them once the linter is no longer run on CI.

hodlinator replied:
Current approach:
* Linter is deleted by another PR - merge conflict has obvious resolution.

Suggested approach:
* Adding `#define`'s to temporarily work around a linter that is about to be removed, only to clean it up in another change seems like extra churn.

Regardless:
* Possible need to call other `tfm::*format*` function based on #30928.

Maybe I'm missing something.

I'm not in a hurry to get this merged before other PRs in this area. Happy move this to draft for now so other PRs can merge and deal with churn myself before undrafting.

hodlinator replied:
(Latest push 071342d4c7627200323d8f22cc37ea022b2e9d69 doesn't touch linter - resolving).

hodlinator commented on src/test/util_string_tests.cpp:
```diff
@@ -13,17 +13,50 @@ BOOST_AUTO_TEST_SUITE(util_string_tests)
+inline void PassFmt(ConstevalFormatString<NumArgs> fmt)
 {
-    // This was already executed at compile-time, but is executed again at run-time to avoid -Wunused.
-    decltype(fmt)::Detail_CheckNumFormatSpecifiers(fmt.fmt);
```
Good point about coverage! Error paths at runtime should be covered by `FailFmtWithError`, but happy path might only be run at compile time. Will bring it back.

hodlinator replied:
Done in 2955b1a1f3eec934960e880963a09b359d828721.

l0rinc commented on src/test/util_string_tests.cpp:
```diff
@@ -13,17 +13,57 @@ BOOST_AUTO_TEST_SUITE(util_string_tests)
     decltype(fmt)::Detail_CheckNumFormatSpecifiers(fmt.fmt);
+
+    // Prove parity with tinyformat
+    switch (NumArgs) {
```
I like the idea of testing the validator together with the implementation!

But I'm not in love with the current approach.

Why are we testing `PassFmt<1>("%02d");` with `foo` - I don't think it helps with understanding how formatter works.

Also `PassFmt<1>("%s")` made sense when we were only validating the number of args, but we've extended it since, I think we should extend examples with the actual parameters, e.g.
* `PassFmt("%s", "test");`
* `PassFmt("%02d", 42);`
* `PassFmt("%12$s %2$s %1$s", "1", "2", "3", "4", "5", "6", "7", "8", "9", "10", "11", "12");`

In which case each example would document the format that we're exercising more specifically.

It would also obviate the parameter count by examples, would all be runtime (seems to me that would simplify a lot) and comparison would simply depend on the behavior of `tfm::format`, e.g. validation would be something like:
```C++
template <typename... Args>
inline bool ValidateFmt(const char* fmt, const char* expected_error, const Args&... args)
{
    try {
        tfm::format(fmt, args...);
        ConstevalFormatString<sizeof...(Args)>::Detail_CheckNumFormatSpecifiers(fmt);
        return true;
    } catch (const tfm::format_error&) {
        using ErrType = const char*;
        auto check_throw = [expected_error](const ErrType& str) { return str == expected_error; };
        BOOST_CHECK_EXCEPTION(ConstevalFormatString<sizeof...(Args)>::Detail_CheckNumFormatSpecifiers(fmt), ErrType, check_throw);
        return false;
    }
}

template <typename... Args>
void PassFmt(const char* fmt, const Args&... args) { BOOST_CHECK(ValidateFmt(fmt, nullptr, args...)); }

template <typename... Args>
void FailFmt(const char* fmt, const char* expected_error, const Args&... args) { BOOST_CHECK(!ValidateFmt(fmt, expected_error, args...)); }
```

and usage would be something like
```
    PassFmt("%%");
    PassFmt("%s", "test");

    auto err_num{"Format specifier count must match the argument count!"};
    FailFmt("%s", err_num);
    FailFmt("%s", err_num, "test", "extra");
```

hodlinator replied:
I like parts of your direction here, and have taken the time to try it out in 2fb2e72f45a1c81d036515861713371fec2bbe2c. My aim is to retain the tests of leaving out one arg though (which your suggestion doesn't include). I'm not enough of a variadic template arg magician to figure out how to skip the last argument, which forces me to still lean on maflcko's `tuple_cat` approach, leaving it feeling somewhat half-baked.

l0rinc replied:
> My aim is to retain the tests of leaving out one arg though (which your suggestion doesn't include)

I didn't include it since I though a single instance of those is enough in the tests.

I would prefer having concrete typed examples over retesting -1 and +1 args (for which explicit examples should likely suffice)

hodlinator replied:
I prefer retaining negative tests to increase certainty and would rather not do the `PassFmt<1>("%s")` -> `PassFmt("%s", "test")` refactor in this PR if that's okay with you.

hodlinator commented on src/test/util_string_tests.cpp:
```diff
@@ -13,17 +13,57 @@ BOOST_AUTO_TEST_SUITE(util_string_tests)
+inline void PassFmt(ConstevalFormatString<NumArgs> fmt)
 {
-    // This was already executed at compile-time, but is executed again at run-time to avoid -Wunused.
+    // Exercise happy paths at run-time for code coverage metrics.
```
`constexpr` functions are not guaranteed to be able to evaluate at compile time for all possible parameters, so exercising that path in the test still provides some value IMO.

hodlinator commented:
(Just added co-authorship of maflcko in latest push from 071342d4c7627200323d8f22cc37ea022b2e9d69 -> a911efc6e868ab1a23d05ee8ab3e94f7ecb89e2b).

hodlinator commented on src/test/util_string_tests.cpp:
```diff
@@ -10,20 +10,42 @@ using namespace util;
+
+    // Prove parity with tinyformat
+    TfmF(fmt.fmt, std::tuple_cat(std::array<int, NumArgs>{}));
+    if constexpr (NumArgs > 0) {
+        BOOST_CHECK_THROW(TfmF(fmt.fmt, std::tuple_cat(std::array<int, NumArgs - 1>{})), tfm::format_error);
+    }
```
Surely that's more than style? :)

Would be nice to have both -1 and +1. Tried out your suggestion but tinyformat fails to throw in 8 of the cases. For example in `"%12$s 999$s %2$s"` with 13 arguments, only using the 12th is not considered an error.

hodlinator commented:
@maflcko
> Are you still working on this?

I'm waiting on #30928 before un-drafting this.

ryanofsky commented on src/test/util_string_tests.cpp:
```diff
@@ -58,9 +80,9 @@ BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
-    PassFmt<1>("%*c");
-    PassFmt<2>("%2$*3$d");
-    PassFmt<1>("%.*f");
+    PassFmtIncorrect<1, 2>("%*c");
```
In commit "test: Prove+document ConstevalFormatString/tinyformat parity" (a911efc6e868ab1a23d05ee8ab3e94f7ecb89e2b)

Maybe drop WrongNumArgs and just pass CorrectArgs? Unless I'm missing something it seems like PassFmtIncorrect could choose the wrong number of args arbitrarily like PassFmt does.

hodlinator replied:
The known cases where I was using `PassFmtIncorrect` have now become valid `PassFmt` thanks to your #31174. `PassFmtIncorrect` is no more.

ryanofsky reviewed (approved):
Code review ACK a911efc6e868ab1a23d05ee8ab3e94f7ecb89e2b. I didn't dig into previous discussion and it seems possible another approach to extending the tests might have advantages over this one. But the code change looks good and seems like an easy way of leveraging existing test cases to check for compatibility with tinyformat.

hodlinator commented on src/test/util_string_tests.cpp:
```diff
@@ -10,13 +10,35 @@ using namespace util;
+    // Disprove parity with tinyformat
+    static_assert(WrongNumArgs != CorrectArgs);
+    TfmF(fmt.fmt, std::tuple_cat(std::array<int, CorrectArgs>{}));
+    BOOST_CHECK_THROW(TfmF(fmt.fmt, std::tuple_cat(std::array<int, WrongNumArgs>{})), tfm::format_error);
```
Thanks! Ended up naming it **Tfm**FormatZeroes in latest push to highlight the tinyformat part being essential.

hodlinator reviewed (commented):
So, I was a bit late reacting to the merge of #31174, but here we are - rebased and ready for review again.

Added a commit regarding non-parity of `"%n"` as suggested in https://github.com/bitcoin/bitcoin/pull/31174#issuecomment-2447224880.

maflcko reviewed (approved):
lgtm, but it would be better to merge the conflicting non-draft pull with two acks first. This will also avoid having to touch some lines twice, which are touched here.

l0rinc commented on src/test/util_string_tests.cpp:
```diff
@@ -11,18 +11,30 @@ using namespace util;
+
+    // Prove parity with tinyformat, which will throw if the number of
+    // conversion specifiers is incorrect.
+    TfmFormatZeroes<NumArgs>(fmt.fmt);
```
I was wondering how it will behave with `%c` which would add a `\0` which could end the const char* prematurely, but since we have a string it seems to work correctly.

Are there any checks we could do with the result here? Or make `TfmFormatZeroes` void if we don't need the actual result. Maybe something like this:
```suggestion
    auto result{TfmFormatZeroes<NumArgs>(fmt.fmt)};
    BOOST_CHECK_EQUAL(result.empty(), std::string_view{fmt.fmt}.empty());
```

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30933#discussion_r1869004667

I think suggested check would be broken in the case of `strprintf("%.0s", 0)` where result should be empty (because it is printing the first 0 characters of a string) even though the format string empty. It has hard to think of a reliable check that could be done here that would also be meaningful.

l0rinc replied:
We can void the method in that case

l0rinc commented on src/test/util_string_tests.cpp:
```diff
@@ -11,18 +11,30 @@ using namespace util;
 BOOST_AUTO_TEST_SUITE(util_string_tests)
 
+template <unsigned NumArgs>
+std::string TfmFormatZeroes(const char* fmt)
```
Is it coincidental that every single format type accepts `0` as a valid substitution?

l0rinc replied:
So %n is the only one that doesn't accept zeroes, right?

hodlinator replied:
Have not fuzzed tinyformat but that seems like a good guess.

ryanofsky commented on src/test/util_string_tests.cpp:
```diff
@@ -11,13 +11,25 @@ using namespace util;
+    // Exercise happy paths at run-time for code coverage metrics.
     decltype(fmt)::Detail_CheckNumFormatSpecifiers(fmt.fmt);
+
+    // Prove parity with tinyformat, which will throw if the number of
+    // conversion specifiers is incorrect.
```
In commit "test: Prove+document ConstevalFormatString/tinyformat parity" (db081bd66015f7899696e6eb4a76052446685b64)

Can there be a test here or somewhere in this file that verifies calls to format really do throw if the number of specifiers is wrong? I am concerned there could be a future change like #30928 where tinyformat errors start being handled differently and these tests become broken without someone noticing.

ryanofsky commented on src/test/util_string_tests.cpp:
```diff
@@ -122,6 +122,12 @@ BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
     FailFmtWithError<2>("%1$.*2$", err_term);
     FailFmtWithError<2>("%1$9.*2$", err_term);
+
+    // Non-parity between tinyformat and ConstevalFormatString
```
In commit "test: Document non-parity between tinyformat and ConstevalFormatstring for "%n"" (2c82ee3219ee186b9d3998a129bff1bc809894d3):

Another case where this is not parity is when a noninteger type is used as a width like `strprintf("%*s", "hi", "hi"),` which throws a `tinyformat::format_error` exception "tinyformat: Cannot convert from argument type to integer for use as variable width or precision". A test could be added for this as well.

ryanofsky reviewed (approved):
Code review ACK 2c82ee3219ee186b9d3998a129bff1bc809894d3

hodlinator committed:
test: Add missing %c character test

Proves tinyformat doesn't trigger an exception for \0 characters.

Co-Authored-By: Lőrinc <pap.lorinc@gmail.com>

ryanofsky commented on src/test/util_string_tests.cpp:
```diff
@@ -11,13 +11,25 @@ using namespace util;
     // Execute compile-time check again at run-time to get code coverage stats
     util::detail::CheckNumFormatSpecifiers<NumArgs>(fmt.fmt);
+
+    // Prove parity with tinyformat, which will throw if the number of
```
In commit "test: Prove+document ConstevalFormatString/tinyformat parity" (4352596f4167d3e0c1d39ee46c6d5096f0228c61)

Not sure this is really parity, since it only shows that tinyformat doesn't throw when ConstevalFormatString doesn't throw, not that tinyformat throws when ConstevalFormatString does (which we probably don't care about). A more literal comment like "Make sure tinyformat doesn't throw if ConstevalFormat didn't throw" might make it more obvious what benefit of this check is.

hodlinator replied:
Updated with more verbose comment, let me know if you think it's okay.

```C++
// If ConstevalFormatString didn't throw above, make sure tinyformat doesn't
// throw either for the same format string and parameter count combination.
// Proves that we have some extent of protection from runtime errors
// (tinyformat may still throw for some type mismatches).
```

ryanofsky reviewed (approved):
Code review ACK 8e7b252d8adb2821b946bd9052f1a707c4edea4f

hodlinator commented on src/test/util_string_tests.cpp:
```diff
@@ -110,6 +122,12 @@ BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
     FailFmtWithError<2>("%1$.*2$", err_term);
     FailFmtWithError<2>("%1$9.*2$", err_term);
+
+    // Belt & suspenders for tinyformat behavior
```
Thanks! Taken.

hodlinator commented on src/test/util_string_tests.cpp:
```diff
@@ -11,13 +11,25 @@ using namespace util;
+
+    // Prove parity with tinyformat, which will throw if the number of
+    // conversion specifiers is incorrect.
+    TfmFormatZeroes<NumArgs>(fmt.fmt);
```
And here I was being proud about having added `BOOST_CHECK_NO_THROW` to your suggested non-parity checks at the bottom. :) Good point, also added to the original `CheckNumFormatSpecifiers`-call in `PassFmt` as I ended up touching that line.

hodlinator commented on src/test/util_string_tests.cpp:
```diff
@@ -123,6 +123,17 @@ BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
+        HasReason{"tinyformat: Cannot convert from argument type to integer for use as variable width or precision"});
+    BOOST_CHECK_NO_THROW(util::detail::CheckNumFormatSpecifiers<2>("%.*s"));
+    BOOST_CHECK_EXCEPTION(tfm::format(std::string{"%.*s"}, "hi", "hi"), tfm::format_error,
+        HasReason{"tinyformat: Cannot convert from argument type to integer for use as variable width or precision"});
```
I think it is valid to have the `CheckNumFormatSpecifiers` calls to prove they work.

In the latest push I've at least removed the `util::detail::` parts to decrease noise which happily makes the format strings line up.

hodlinator reviewed (commented):
Latest push:
- Adds `using util::detail::CheckNumFormatSpecifiers;`, decreasing noise in the rest of the changes.
- Adds `BOOST_CHECK_NO_THROW` to both `CheckNumFormatSpecifiers` and tinyformat calls in `PassFmt`.
- Makes comments more descriptive/useful.
- Adds context to commit message of the last commit.

Also updated PR summary:
"Makes unequivocally clear the extent of parity." -> "Clarifies and puts the extent of parity under test."

l0rinc commented on src/test/util_string_tests.cpp:
```diff
@@ -8,21 +8,22 @@
 {
     // Execute compile-time check again at run-time to get code coverage stats
-    util::detail::CheckNumFormatSpecifiers<NumArgs>(fmt.fmt);
+    BOOST_CHECK_NO_THROW(CheckNumFormatSpecifiers<NumArgs>(fmt.fmt));
```
👍 for `BOOST_CHECK_NO_THROW`, documents the reason for the dangling call a lot better

l0rinc commented on src/test/util_string_tests.cpp:
```diff
@@ -8,21 +8,22 @@
 // args directly. Normally PassFmt<sizeof...(Args)> would be used.
 template <unsigned NumArgs>
-inline void PassFmt(util::ConstevalFormatString<NumArgs> fmt)
+inline void PassFmt(ConstevalFormatString<NumArgs> fmt)
```
nit: given that `PassFmt` is local to `util_string_tests`, I think `static` would make more sense here (and the other helpers which sometimes have inline and sometimes nothing):
```suggestion
static void PassFmt(ConstevalFormatString<NumArgs> fmt)
```

hodlinator replied:
They are also template functions, so implicitly inline, and arguably local already. Holding off on this for now.

l0rinc replied:
I find it inconsistent that `PassFmt` and `FailFmtWithError` are marked inline, but the new `TfmFormatZeroes` is not.
I'm fine with not making them static, but we could at least remove the redundant `inline`s - not critical, I won't block, would just prefer consistency.

l0rinc commented on src/test/util_string_tests.cpp:
```diff
@@ -110,6 +126,26 @@ BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
+    BOOST_CHECK_EXCEPTION(tfm::format(std::string{"%.*s"}, "hi", "hi"), tfm::format_error,
+        HasReason{"tinyformat: Cannot convert from argument type to integer for use as variable width or precision"});
+
+    // Ensure that tinyformat will throws if format string contains wrong number
```
```suggestion
    // Ensure that tinyformat will throw if format string contains wrong number
```

maflcko reviewed (approved):
nothing blocking, just a nit.


re-ACK 2951532a2ca09ac8ebdea066fa20826fdb6fa8e5 💼

<details><summary>Show signature</summary>

Signature:

```
untrusted comment: signature from minisign secret key on empty file; verify via: minisign -Vm "${path_to_any_empty_file}" -P RWTRmVTMeKV5noAMqVlsMugDDCyyTSbA3Re5AkUrhvLVln0tSaFWglOw -x "${path_to_this_whole_four_line_signature_blob}"
RUTRmVTMeKV5npGrKx1nqXCw5zeVHdtdYURB/KlyA/LMFgpNCs+SkW9a8N95d+U4AP1RJMi+krxU1A3Yux4bpwZNLvVBKy0wLgM=
trusted comment: re-ACK 2951532a2ca09ac8ebdea066fa20826fdb6fa8e5 💼
LJt7P3harWblu2dLt4t2t8Z3yPelpL/kKq/KB7AoO45V3vTgMr8HXd6YbZiCCx5Y3BTDTIlVohJmss9B0+l9DA==
```

</details>

l0rinc commented on src/test/util_string_tests.cpp:
```diff
@@ -111,6 +125,15 @@ BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
+    // formats the strings, and will need to be updated if tinyformat is changed
+    // not to throw on failure.
+    BOOST_CHECK_EXCEPTION(TfmFormatZeroes<2>("%s"), tfm::format_error,
+        HasReason{"tinyformat: Not enough conversion specifiers in format string"});
```
nit: don't feel strongly about it, just noting that `HasReason` can do partial matches, we can also check something shorter like:
```suggestion
        HasReason{"Not enough conversion specifiers"});
```
(which would enable this being a one-liner)

hodlinator replied:
Thanks for the suggestion, it's a trade-off, holding off on changing this.

l0rinc commented on src/test/util_string_tests.cpp:
commit message typo:
```
- For "%n", which is supposed to write to the argument for printf.
 ```

hodlinator replied:
Cheers, taken.

l0rinc commented on src/test/util_string_tests.cpp:
```diff
@@ -8,28 +8,44 @@
 
 BOOST_AUTO_TEST_SUITE(util_string_tests)
 
+template <unsigned NumArgs>
+void TfmFormatZeroes(const char* fmt)
+{
+    std::apply([fmt](auto... args) {
+        (void)tfm::format(std::string{fmt}, args...);
+    }, std::tuple_cat(std::array<int, NumArgs>{}));
+}
```
nit: we could avoid `[fmt]` via `[&]` here, make it `static` and give it a `std::string` param directly:
```suggestion
template <unsigned NumArgs>
static void TfmFormatZeroes(const std::string& fmt)
{
    std::apply([&](auto... args) { tfm::format(fmt, args...); }, std::array<int, NumArgs>{});
}
```

hodlinator replied:
Thanks, taken in broad strokes.

Temporarily dded
```diff
BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
{
+    TfmFormatZeroes<10>("blah")
```
together with
```diff
void TfmFormatZeroes(const std::string& fmt)
{
    std::apply([&](auto... args) {
+        fprintf(stderr, "ARG COUNT: %d", sizeof...(args));
+        assert(false);
        (void)tfm::format(fmt, args...);
    }, std::array<int, NumArgs>{});
}
```
To verify that `tuple_cat` was unnecessary.

l0rinc commented on test/lint/run-lint-format-strings.py:
```diff
@@ -15,6 +15,8 @@
 FALSE_POSITIVES = [
     ("src/clientversion.cpp", "strprintf(_(COPYRIGHT_HOLDERS), COPYRIGHT_HOLDERS_SUBSTITUTION)"),
     ("src/test/translation_tests.cpp", "strprintf(format, arg)"),
+    ("src/test/util_string_tests.cpp", 'tfm::format(std::string{"%*s"}, "hi", "hi")'),
```
What's the added value for extending this linter as well?

hodlinator replied:
Was getting CI failures, so added these exceptions. I considered extending the linter implementation to handle them correctly, but part of the raison d'étre of `ConstevalFormatString` is getting further towards being able to remove the linter in the first place.

l0rinc replied:
Ah like these new ones you mean?
```
[20:48:24.756] src/test/util_string_tests.cpp: Expected 0 argument(s) after format string but found 1 argument(s): tfm::format(ConstevalFormatString<2>{"%*s"}, "hi", "hi")
[20:48:24.756] src/test/util_string_tests.cpp: Expected 0 argument(s) after format string but found 1 argument(s): tfm::format(ConstevalFormatString<2>{"%.*s"}, "hi", "hi")
```

hodlinator replied:
Exactly :D

Fixed in latest push.

@maflcko I realize you were concerned about touching this linter in https://github.com/bitcoin/bitcoin/pull/30933#discussion_r1768048796, do you have an alternate approach suggestion?

ryanofsky reviewed (approved):
Code review ACK 2951532a2ca09ac8ebdea066fa20826fdb6fa8e5 just adding using declaration and improving comments since last reviewe

hodlinator committed:
test: Prove+document ConstevalFormatString/tinyformat parity

Co-Authored-By: Lőrinc <pap.lorinc@gmail.com>
Co-Authored-By: MarcoFalke <*~=`'#}+{/-|&$^_@721217.xyz>
Co-Authored-By: Ryan Ofsky <ryan@ofsky.org>

l0rinc commented:
ACK 8ee014020071a60d1bdcdd6c16a7e515c6576fea

hodlinator committed:
refactor test: Profit from using namespace + using detail function

Also adds BOOST_CHECK_NO_THROW() while touching that line, clarifying part of what we are checking for.

Also removed redundant inline from template functions in .cpp file.

hodlinator committed:
test: Document non-parity between tinyformat and ConstevalFormatstring

- For "%n", which is supposed to write to the argument for printf.
- For string/integer mismatches of width/precision specifiers.

Co-Authored-By: Ryan Ofsky <ryan@ofsky.org>

hodlinator reviewed (commented):
Had to re-push after fully grokking https://github.com/bitcoin/bitcoin/pull/30933#discussion_r1871989800.

l0rinc commented on src/test/util_string_tests.cpp:
```diff
@@ -110,6 +126,24 @@ BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
+    // tinyformat throws but ConstevalFormatString does not.
+    BOOST_CHECK_EXCEPTION(tfm::format(ConstevalFormatString<1>{"%n"}, 0), tfm::format_error,
+        HasReason{"tinyformat: %n conversion spec not supported"});
+    BOOST_CHECK_EXCEPTION(tfm::format(ConstevalFormatString<2>{"%*s"}, "hi", "hi"), tfm::format_error,
```
nit: personally I'm not a fan of "funny" variable names - can we do "X" or "abc" or "text" or something instead? Not a blocker, just a preference

hodlinator replied:
Good point. Might go with something like this (slightly useful) if I re-touch:
```suggestion
    BOOST_CHECK_EXCEPTION(tfm::format(ConstevalFormatString<2>{"%*s"}, "one", "two"), tfm::format_error,
```

l0rinc commented on src/test/util_string_tests.cpp:
```diff
@@ -110,6 +126,24 @@ BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
+
+    // Non-parity between tinyformat and ConstevalFormatString.
+    // tinyformat throws but ConstevalFormatString does not.
+    BOOST_CHECK_EXCEPTION(tfm::format(ConstevalFormatString<1>{"%n"}, 0), tfm::format_error,
```
these are basically TODOs for follow-up PRs, right?
Can we add `ConstevalFormatString` parity check for the [FUZZ tests](https://github.com/bitcoin/bitcoin/blob/master/src/test/fuzz/strprintf.cpp#L17) as well (guess that should reveal other cases where it's misaligned)

hodlinator replied:
I guess a version of `ConstevalFormatString/CheckNumFormatSpecifiers()` could be checked in a test together with `tfm::format`, but I'm not sure it would be a fuzz-test, as `tfm::format()` takes args specified at compile time.

It could conceivably be a pretty advanced fuzz-test that generated C++ code and compiled it as part of the test, not sure if we have that class of tests already or if this would be new, and if it would be accepted.

l0rinc replied:
Why not validate it during runtime via `CheckNumFormatSpecifiers`?

hodlinator replied:
`CheckNumFormatSpecifiers` is currently a template, but sure, could be changed to take the number of specifiers at runtime instead.

If we want to check *parity* in a fuzzing context we should probably check `tfm::format()` as well, with the same input?

hodlinator replied:
Didn't think to lock-down the number of args in that way and still get value from fuzzing. Turned out great!

Are you going to open a PR for this or what do you suggest as next step?

l0rinc commented on src/test/util_string_tests.cpp:
```diff
@@ -110,6 +126,24 @@ BOOST_AUTO_TEST_CASE(ConstevalFormatString_NumSpec)
+        HasReason{"tinyformat: %n conversion spec not supported"});
+    BOOST_CHECK_EXCEPTION(tfm::format(ConstevalFormatString<2>{"%*s"}, "hi", "hi"), tfm::format_error,
+        HasReason{"tinyformat: Cannot convert from argument type to integer for use as variable width or precision"});
+    BOOST_CHECK_EXCEPTION(tfm::format(ConstevalFormatString<2>{"%.*s"}, "hi", "hi"), tfm::format_error,
```
nit: `strprintf` is a bit shorter than `tfm::format`

hodlinator replied:
Shorter yes, but this is more direct/concrete.

l0rinc reviewed (approved):
ACK c93bf0e6e2cd0e9dc979ed996e41cc1034bc1ad1
[Changes](https://github.com/bitcoin/bitcoin/compare/8ee014020071a60d1bdcdd6c16a7e515c6576fea..c93bf0e6e2cd0e9dc979ed996e41cc1034bc1ad1):
* Unified templates
* Adjusted non-parity of `tfm::format` of `ConstevalFormatString`

Would be curious if we could extend this to fuzzing as well - but it can be explored in a different PR as well

ryanofsky reviewed (approved):
Code review ACK c93bf0e6e2cd0e9dc979ed996e41cc1034bc1ad1. Just a few cleanups tweaking function declarations and commit comments and consolidating some test cases since last review.
